addons/libs/dialog.lua

```diff
@@ -0,0 +1,158 @@
+-- While the IDs can be collected in-game, they occasionally
+-- change and would otherwise need to be manually updated.
+-- It can also be used to find and decode an entry given the ID.
+
+-- Common parameters:
+-- dat: The entire content of the zone dialog DAT file
+-- i.e. local dat = io.open('path/to/dialog/DAT', 'rb'):read('*a')
+-- entry: The string you are looking for. If you do not know the
+-- entire string, use dev_find_substring.
+
+local xor = require('bit').bxor
+require('pack')
+local string = require('string')
+local unpack = string.unpack
+local pack = string.pack
+local find = string.find
+
+local dialog = {}
+
+local function decode(int)
+    return xor(int, 0x80808080)
+end
+local encode = decode
+
+local floor = require('math').floor
+local function binary_search(pos, dat, n)
+    local l, r, m = 1, n
+    while l < r do
+        m = floor((l+r)/2)
+        if decode(unpack('<I', dat, 1+4*m)) < pos then
+        -- offset given by mth ID < offset to string
+            l = m + 1
+        else
+            r = m
+        end
+    end
+    return l-2 -- we want the index to the left of where "pos" would be placed
+end
+
+local function plain_text_gmatch(text, substring, n)
+    n = n or 1
+    return function()
+        local pos = find(text, substring, n, true)
+        if pos then n = pos + 1 end
+        return pos
+    end
+end
+
+function dialog.dev_get_offset(dat, id) -- sanity check function
+    return string.format('%x', decode(unpack('<I', dat, 5+4*id)))
+end
+
+-- Returns the number of entries in the given dialog DAT file
+function dialog.entry_count(dat)
+    return decode(unpack('<I', dat, 5))/4
+end
+
+-- Returns an array-like table containing every ID which matched
+-- the given entry.
+-- An important note about this function: The tables contain an
+-- enormous number of duplicate entries. Additionally, some short
+-- entries can occur as substrings of longer entries. Results
+-- which fall into the second case should be filtered out.
+-- If you are dealing with the first case, it may help to look at
+-- any other messages which were received at the same time.
+function dialog.get_ids_matching_entry(dat, entry)
+    local last_offset = decode(unpack('<I', dat, 5))
+    local res = {}
+    local n = 0
+    local start = 5
+    for i in plain_text_gmatch(dat, entry, last_offset) do
+        local encoded_pos = pack('<I', encode(i-5))
+        local offset = find(dat, encoded_pos, start, true)
+        if offset then
+            offset = offset-1
+            local next_pos
+            if offset > last_offset then
+                break
+            elseif offset == last_offset then
+                next_pos = #dat+1
+            else
+                next_pos = decode(unpack('<I', dat, offset+5))+5
+            end
+
+            if next_pos-i == #entry then
+                n = n + 1
+                res[n] = (offset-4)/4
+            end
+            start = offset+1
+        end
+    end
+    res.n = n
```

Comment:
Not sure why these result tables need to contain a length variable. Unless you meant to simulate lists, but then I'd just use lists. Same below.

Reply:
I just tossed it in since it was already available. Otherwise, you'd end up calculating it anyway.

Reply:
It's a quick way to tell if the message is unique. If `n ~= 1` you have more work to do before you can be certain you've found the correct ID.
I guess I don't really need it for the one use I had, though.

Reply:
I might not be expressing myself properly, I just meant that `#` is used to determine length. Using a custom table key for that feels weird to me. The `lists` library does that, but that was never meant to be publicly used, if Lua supported private variables, it would have been private. It feels very unintuitive, and I don't think anyone would use it over `#` (which would still work).

Reply:
Maybe I shouldn't admit this now, but `list.n` became so engrained in my programming that I use it everywhere now. Outside of Windower, and even in some other languages.